### PR TITLE
Coverage thresholds checks

### PR DESCRIPTION
**Pull request by vladsaveliev**
To address #2122 - rerun mosdepth if the coverage thresholds file is missing; if mosdepth is older, just skip processing thresholds.

### CODE REVIEW

File: bcbio/variation/coverage.py [hunk 1]
```diff
@@ -208,12 +207,17 @@ def run_mosdepth(data, target_name, bed_file, per_base=False, quantize=None, thr
     bam_file = dd.get_align_bam(data) or dd.get_work_bam(data)
     work_dir = utils.safe_makedir(os.path.join(dd.get_work_dir(data), "coverage", dd.get_sample_name(data)))
     prefix = os.path.join(work_dir, "%s-%s" % (dd.get_sample_name(data), target_name))
+    thresholds = thresholds and LooseVersion(check_output(["mosdepth", "--version"]).split()[1].strip()) >= LooseVersion("0.2.0")
```
Comment (chapmanb) [inline bcbio/variation/coverage.py]: We can assume that the user will have an up to date mosdepth so always apply the thresholds argument. The case we want to catch is when mosdepth has already been run with an older version (so won't have the thresholds file).

File: bcbio/variation/coverage.py [hunk 2]
```diff
@@ -208,12 +207,17 @@ def run_mosdepth(data, target_name, bed_file, per_base=False, quantize=None, thr
...
                       ("%s.per-base.bed.gz" % prefix) if per_base else None,
                       ("%s.regions.bed.gz" % prefix) if bed_file else None,
                       ("%s.quantized.bed.gz" % prefix) if quantize else None,
                       ("%s.thresholds.bed.gz" % prefix) if thresholds else None)
-    if not utils.file_uptodate(out.dist, bam_file):
+    if not utils.file_uptodate(out.dist, bam_file) \
+            or (per_base and not utils.file_uptodate(out.per_base, bam_file)) \
+            or (bed_file and not utils.file_uptodate(out.regions, bam_file)) \
+            or (quantize and not utils.file_uptodate(out.quantize, bam_file)) \
+            or (thresholds and not utils.file_uptodate(out.thresholds, bam_file)):
```
Comment (chapmanb) [inline bcbio/variation/coverage.py]: I'd skip all these extra checks here. If the user has run it previously we don't want to re-run, but later just skip if the thresholds file isn't present.

File: bcbio/variation/coverage.py [hunk 3]
```diff
@@ -226,9 +230,10 @@ def run_mosdepth(data, target_name, bed_file, per_base=False, quantize=None, thr
                 quant_export += " && "
             else:
                 quant_arg, quant_export = "", ""
-            thresholds = "-T " + ",".join([str(t) for t in thresholds]) if thresholds else ""
+
+            thresholds_cmdl = "-T " + ",".join([str(t) for t in thresholds]) if out.thresholds else ""
```
Comment (chapmanb) [inline bcbio/variation/coverage.py]: This is causing the test failure (https://travis-ci.org/chapmanb/bcbio-nextgen/builds/294676740#L4918) but we can revert back have this as long as we have thresholds set (assume mosdepth 0.2.0 or better installed).

File: bcbio/variation/coverage.py [hunk 4]
```diff
@@ -252,12 +257,12 @@ def coverage_region_detailed_stats(target_name, bed_file, data, out_dir):
         ready_depth = tz.get_in(["depth", target_name], data)
         cov_file = ready_depth["regions"]
         dist_file = ready_depth["dist"]
-        thresholds_file = ready_depth["thresholds"]
+        thresholds_file = ready_depth.get("thresholds")
```
Comment (chapmanb) [inline bcbio/variation/coverage.py]: If you add an existence check here and set thresholds_file to None if missing I think that should capture what we need.

**Comment by chapmanb** [comment]
Vlad -- thanks much for tackling this issue (from #2122 discussion). I left some throughts: my approach would be to always assume mosdepth 0.2.0 or greater if running from scratch and then check afterwards and only return the thresholds file if present. Thanks again for looking at this.

**Comment by vladsaveliev** [comment]
Passing now!

**Comment by chapmanb** [comment]
Vlad -- thanks for all this work, much appreciated.

**Pull request merged**